File: Daily/198. House Robber/Solution1.cpp
#include <iostream>
#include <vector>
#include <functional>
using namespace std;
class Solution {
    public:
        int rob(vector<int>& nums) {
            /*
                Optimized Solution (Memomization)
                Time Complexity = O(n)
                Space Complexity = O(n)
            */
            vector<int> cache(nums.size(), -1);
            function<int(int)> find_Max_Amount = [&](int index) {
                if(index >= nums.size()) return 0;
                if(cache[index] == -1) cache[index] = max(nums[index] + find_Max_Amount(index + 2), find_Max_Amount(index + 1));
                return cache[index];
            };
            return find_Max_Amount(0);
        }
};
int main() {
   int n; cin >> n;
   vector<int> nums(n, 0); for(int index = 0; index < n; ++index) cin >> nums[index];
   Solution sol; cout << sol.rob(nums) << endl; 
}

File: Daily/198. House Robber/Solution1.py
from typing import List
class Solution:
    def rob(self, nums: List[int]) -> int:
        '''
            Optimized Solution (Memomization)
            Time Complexity = O(n)
            Space Complexity = O(n)
        '''
        cache = [-1]*len(nums)
        def find_Max_Amount(index: int = 0) -> int:
            if index >= len(nums): return 0
            if cache[index] == -1: 
                cache[index] = max(nums[index] + find_Max_Amount(index + 2), find_Max_Amount(index + 1))
            return cache[index]
        return find_Max_Amount()
if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    print(sol.rob(nums))